Source/ARTRealtime.m

```diff
@@ -814,46 +827,41 @@ - (void)onHeartbeat {
+}
+
 - (void)onConnected:(ARTProtocolMessage *)message {
     _renewingToken = false;
-    
-    // Resuming
-    if (_resuming) {
-        if (![message.connectionId isEqualToString:self.connection.id_nosync]) { // RTN15c3
-            [self.logger warn:@"RT:%p connection \"%@\" has reconnected, but resume failed. Reattaching any attached channels", self, message.connectionId];
-            // Reattach all channels
-            for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
-                ARTAttachRequestMetadata *const metadata = [[ARTAttachRequestMetadata alloc] initWithReason:message.error];
-                [channel reattachWithMetadata:metadata];
-            }
-            _resuming = false;
-        }
-        else if (message.error) {
-            [self.logger warn:@"RT:%p connection \"%@\" has resumed with non-fatal error \"%@\"", self, message.connectionId, message.error.message];
-            // The error will be emitted on `transition`
-        }
-        else {
-            [self.logger debug:@"RT:%p connection \"%@\" has reconnected and resumed successfully", self, message.connectionId];
-        }
-    }
-    
+
     switch (self.connection.state_nosync) {
         case ARTRealtimeConnecting: {
-            // If there's no previous connectionId, then don't reset the msgSerial
-            //as it may have been set by recover data (unless the recover failed).
-            NSString *prevConnId = self.connection.id_nosync;
-            BOOL connIdChanged = prevConnId && ![message.connectionId isEqualToString:prevConnId];
-            BOOL recoverFailure = !prevConnId && message.error;
-            if (connIdChanged || recoverFailure) {
-                [self.logger debug:@"RT:%p msgSerial of connection \"%@\" has been reset", self, self.connection.id_nosync];
-                self.msgSerial = 0;
-                self.pendingMessageStartSerial = 0;
+            ARTResumeRequestResponse *response = [[ARTResumeRequestResponse alloc] initWithCurrentConnectionID:_connection.id_nosync
+                                                                                               protocolMessage:message
+                                                                                                  errorChecker:[[ARTDefaultErrorChecker alloc] init]];
+            if (response.type == ARTResumeRequestResponseTypeValid) {
+                [self.logger debug:@"RT:%p connection \"%@\" has reconnected and resumed successfully", self, message.connectionId];
+            }
+            else if (response.type == ARTResumeRequestResponseTypeInvalid) {
+                [self.logger warn:@"RT:%p connection \"%@\" has resumed with non-fatal error \"%@\"", self, message.connectionId, message.error.message];
+                [self resetSerials];
+            }
+            else if (response.type == ARTResumeRequestResponseTypeFatalError || response.type == ARTResumeRequestResponseTypeTokenError) {
```

Comment:
Why not a `switch` here?

Reply:
Good point, but I'm rewriting this piece anyway.